### PR TITLE
Multiple Features

### PR DESCRIPTION
**Pull request by protich**
**Upgrader and db migrator**
The upgrader now supports ability to monkey patch the database.

**Unified ticket thread table**
Messages, Response and Notes stored in one table.

**Ticket events tracking** 
Track ticket events like Opened, Closed, Reopened, Overdue...etc

**Advanced Ticket Search**

### CODE REVIEW

**Comment by greezybacon** [comment]
Advanced search works, database upgrader almost works. Please log errors to a log file for system admin review.
- Try a fresh install
- Run the lint test suite against the new stuff

File: setup/inc/class.upgrader.php [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+            if($task['status']) //Progress report... 
+                $action.=' ('.$task['status'].')';
+        } elseif($this->isUpgradable() && ($nextversion = $this->getNextVersion())) {
+            $action = "Upgrade to $nextversion";
+        }
+
+        return $action;
+    }
+
+    function getNextStepInfo() {
```
Comment (greezybacon) [inline setup/inc/class.upgrader.php]: This function is unused. Remove it

File: include/ajax.tickets.php [hunk 2]
```diff
@@ -84,6 +84,101 @@ function searchByEmail() {
...
+            
+        //dates
+        $startTime  =($_REQUEST['startDate'] && (strlen($_REQUEST['startDate'])>=8))?strtotime($_REQUEST['startDate']):0;
+        $endTime    =($_REQUEST['endDate'] && (strlen($_REQUEST['endDate'])>=8))?strtotime($_REQUEST['endDate']):0;
+        if( ($startTime && $startTime>time()) or ($startTime>$endTime && $endTime>0))
+            $startTime=$endTime=0;
+        
+        //Have fun with dates.
+        if($startTime)
+            $qwhere.=' AND ticket.created>=FROM_UNIXTIME('.$startTime.')';
```
Comment (greezybacon) [inline include/ajax.tickets.php]: Think you meant `$where .= ...`

**Comment by greezybacon** [comment]
Once merged, create a new pull-request to add an 'annulled' column to the %ticket_event table and, when `closed` and `reopened` events are logged, annul the previous, opposite event if any. For instance, reopening a ticket should annul a previous `closed` event and visa-versa. This will greatly assist the reporting engine generate better statistics.

File: include/ajax.tickets.php (hunk 2 above)
Comment (protich) [inline include/ajax.tickets.php]: Fixed.

File: setup/inc/class.upgrader.php (hunk 1 above)
Comment (protich) [inline setup/inc/class.upgrader.php]: Removed.

**Comment by protich** [comment]
**New install works** : Installer completed with no errors - initial ticket got created as expected.
**Inline Comments** : Commit for the 2 inline comments pushed to the PR branch.

**Pull request merged**